### PR TITLE
solution

### PR DESCRIPTION
**Pull request by ihorandrianov**

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  public app: express.Application;
+  public port: number;
+
+  constructor(controllers, port) {
+    this.app = express();
+    this.port = port;
+
+    this.initializeMiddlewares();
+    this.initializeControllers(controllers);
+    this.initializeErrorHandling();
```
Comment (Andr1yK) [inline src/app.ts:15]: u can create `init` function😃

File: src/db/db.instance.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  private idCounter: number;
+
+  constructor(items = []) {
+    this.items = items;
+    this.idCounter = 0;
+  }
+
+  private getItemIndex(id: number): number {
+    const index = this.items.findIndex((item) => item.id === id);
+    return index;
```
Comment (Andr1yK) [inline src/db/db.instance.ts:12]:
```suggestion
    return this.items.findIndex((item) => item.id === id);
```

File: src/db/db.instance.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
+class DbInstance {
+  private items: any[];
+  private idCounter: number;
+
+  constructor(items = []) {
+    this.items = items;
+    this.idCounter = 0;
+  }
+
+  private getItemIndex(id: number): number {
```
Comment (Andr1yK) [inline src/db/db.instance.ts:10]:
```suggestion
  private getIndex(id: number): number {
```

File: src/db/db.instance.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    const index = this.items.findIndex((item) => item.id === id);
+    return index;
+  }
+
+  public getAll() {
+    return this.items;
+  }
+
+  public createNew<T>(payload: T) {
```
Comment (Andr1yK) [inline src/db/db.instance.ts:19]: As I think, we always create a new instance :)

```suggestion
  public create<T>(payload: T) {
```

File: src/expanses/expanses.controller.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+class ExpansesController {
+  public path = "/expanses";
+  public router = express.Router();
+
+  private expanses;
+
+  constructor() {
+    this.initializeRoutes();
+    this.expanses = new DbInstance();
+  }
```
Comment (Andr1yK) [inline src/expanses/expanses.controller.ts:17]: Let's create a basic (abstract) class `Controller` (you can add `initializeRoutes` method to it as required)

image

File: src/expanses/expanses.interface.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+interface Expanse {
```
Comment (Andr1yK) [inline src/expanses/expanses.interface.ts:1]: maybe u can use `CreateExpanseDto` as a type🤔

File: src/middleware/error.middleware.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { NextFunction, Request, Response } from "express";
+import HttpException from "exceptions/HttpException";
+
+function errorMiddleware(
+  error: HttpException,
+  _request: Request,
+  response: Response,
+  _next: NextFunction
```
Comment (Andr1yK) [inline src/middleware/error.middleware.ts:8]: do u need `next` here?)

File: src/middleware/validation.middleware.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import { validate, ValidationError } from "class-validator";
+import { plainToClass } from "class-transformer";
+import HttpException from "../exceptions/HttpException";
+
+function validationMiddleware(
+  type: any,
+  skipMissingProperties = false
+): RequestHandler {
+  return (req, _res, next) => {
+    validate(plainToClass(type, req.body), {
```
Comment (Andr1yK) [inline src/middleware/validation.middleware.ts:11]: I would use `async wait` :)

File: src/users/users.controller.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    const id = request.params.id;
+    const user = this.users.getById(Number(id));
+    if (user) {
+      response.send(user);
+      return;
+    }
+    next(new UserNotFoundException(id));
+  };
+
+  private editUser = (
```
Comment (Andr1yK) [inline src/users/users.controller.ts:65]: as for the method, it would be more correct to call it `update`

**Review by Andr1yK — CHANGES_REQUESTED** [review]

File: src/expanses/expanses.interface.ts (hunk 6 above)
Comment (ihorandrianov) [reply src/expanses/expanses.interface.ts:1]: CreateExpanse should be without **id** (will fix it in next commit) so i can validate incoming JSON to not contain any external keys, and to not allow duplicate id which is generated on "server"

Comment (ihorandrianov) [reply src/expanses/expanses.interface.ts:1]: Same for User interface and User Dto

Comment (Andr1yK) [reply src/expanses/expanses.interface.ts:1]: u can use something looks like this 

```
type Expanse = CreateExpanseDto & {
  id: number
};

Comment (Andr1yK) [reply src/expanses/expanses.interface.ts:1]: or create a utility for this issue

type Expanse = ModelWithId<CreateExpanseDto>

File: src/users/users.controller.ts [hunk 10]
```diff
@@ -5,18 +5,20 @@ import validationMiddleware from "../middleware/validation.middleware";
...
-  private users;
+  private users: DbInstance;
 
   constructor() {
+    super();
     this.initializeRoutes();
     this.users = new DbInstance();
   }
 
-  private initializeRoutes() {
```
Comment (Andr1yK) [inline src/users/users.controller.ts]: u can make it private :)

File: src/users/users.interface.ts [hunk 11]
```diff
@@ -1,6 +1,7 @@
-interface User {
+import CreateUserDto from "./users.dto";
+
+interface User extends CreateUserDto {
```
Comment (Andr1yK) [inline src/users/users.interface.ts:3]: :fire:

**Review by Andr1yK — APPROVED** [review]

File: src/users/users.controller.ts (hunk 10 above)
Comment (ihorandrianov) [reply src/users/users.controller.ts]: abstract function cant be private compiler says

**Pull request closed without merging**